File: PetMedicalCare/SoftPetBusiness/src/test/java/pe/edu/pucp/bo/ProductoBoTest.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/UnitTests/JUnit5TestClass.java to edit this template
 */
package pe.edu.pucp.bo;

import java.util.ArrayList;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import pe.edu.pucp.softpet.dto.productos.ProductoDto;
import pe.edu.pucp.softpet.dto.productos.TipoProductoDto;

/**
 *
 * @author marti
 */
public class ProductoBoTest {

    private ProductoBo productoBo;

    public ProductoBoTest() {
        this.productoBo = new ProductoBo();
    }

    /**
     * Test of insertar method, of class ProductoBo.
     */
    @Test
    public void testInsertar() {
        System.out.println("=== Test: insertar producto ===");

        // Datos de entrada (nombre y presentacion con espacios/minúsculas)
        String nombre = "   shampoo para perros  ";
        String presentacion = "  botella 250ml  ";
        double precio = 19.90;
        boolean activo = true;

        int tipoProductoId = 1;

        //insertar
        Integer idGenerado = productoBo.insertar(nombre, presentacion, precio, activo, tipoProductoId);

        // Solo verificar que el ID generado sea válido
        assertTrue(idGenerado > 0, "El ID generado debe ser mayor que 0");
    }

    /**
     * Test of modificar method, of class ProductoBO.
     */
    @Test
    public void testModificar() {
        System.out.println("=== Test: modificar producto existente ===");

        int idProductoExistente = 8; // Asumimos que ya existe
        String nuevoNombre = "Shampoo Premium para perros";
        String nuevaPresentacion = "botella 300ml";
        double nuevoPrecio = 25.50;
        boolean nuevoActivo = false;
        int tipoProductoId = 1;

        Integer resultado = productoBo.modificar(
                idProductoExistente,
                nuevoNombre,
                nuevaPresentacion,
                nuevoPrecio,
                nuevoActivo,
                tipoProductoId
        );

        assertTrue(resultado > 0, "El método modificar debe retornar true si la actualización fue exitosa");
    }

    /**
     * Test of eliminar method, of class ProductoBo.
     */
    @Test
    public void testEliminar() {
        System.out.println("=== Test: eliminar producto ===");

        // Insertamos un producto temporal para eliminarlo
        String nombre = "Collar para gatos";
        String presentacion = "Paquete individual";
        double precio = 10.50;
        boolean activo = true;
        int tipoProductoId = 1;

        Integer idGenerado = productoBo.insertar(nombre, presentacion, precio, activo, tipoProductoId);
        assertTrue(idGenerado > 0, "No se pudo insertar el producto para eliminar.");

        // Ahora eliminamos
        Integer resultado = productoBo.eliminar(idGenerado);
        assertTrue(resultado > 0, "El método eliminar debe retornar true si la eliminación fue exitosa");
    }

    /**
     * Test of obtenerPorId method, of class ProductoBo.
     */
    @Test
    public void testObtenerPorId() {
        System.out.println("=== Test: obtener producto por ID ===");

        int idProductoExistente = 1; // Asegúrate de que este ID exista
        ProductoDto producto = productoBo.obtenerPorId(idProductoExistente);

        assertNotNull(producto, "El producto no debe ser null");
        assertEquals(idProductoExistente, producto.getProductoId(), "El ID obtenido no coincide");

    }

    /**
     * Test of listarTodos method, of class ProductoBo.
     */
    @Test
    public void testListarTodos() {
        System.out.println("=== Test: listar todos los productos ===");

        ArrayList<ProductoDto> lista = productoBo.listarTodos();

        // Validaciones básicas
        assertNotNull(lista, "La lista no debe ser null");
        assertFalse(lista.isEmpty(), "La lista no debe estar vacía");

        // Mostrar en consola los productos obtenidos
        System.out.println("Productos encontrados: " + lista.size());
        for (ProductoDto p : lista) {
            System.out.println("-----------------------------------");
            System.out.println("ID: " + p.getProductoId());
            System.out.println("Nombre: " + p.getNombre());
            System.out.println("Presentación: " + p.getPresentacion());
            System.out.println("Precio: " + p.getPrecioUnitario());
            System.out.println("Activo: " + (p.getActivo()? "Sí" : "No"));
            if (p.getTipoProducto() != null) {
                System.out.println("Tipo de producto ID: " + p.getTipoProducto().getTipoProductoId());
            }
        }
    }

}

File: Mis últimos pasos/SoftPetWA/BoTest/SoapTest/GestionLogistica/ProductoBoTest.cs
//using Microsoft.VisualStudio.TestTools.UnitTesting;
//using SoftPetBussiness.ProductoClient;
//using System;
//using System.Collections.Generic;

//namespace SoftPetBusiness.Tests
//{
//    [TestClass]
//    public class ProductoBOTest
//    {
//        private ProductoBO productoBO;
//        private int id;

//        [TestInitialize]
//        public void Setup()
//        {
//            productoBO = new ProductoBO();
//        }

//        [TestMethod]
//        public void TestListarGenericos()
//        {
//            // Obtener la lista de todos los productos
//            List<productoDto> productos = productoBO.ListarTodos();

//            // Verificar que la lista no esté vacía
//            Assert.IsNotNull(productos);
//            Assert.IsTrue(productos.Count > 0, "La lista de productos no debería estar vacía.");

//            List<productoDto> productosActivos = productoBO.ListarActivos();
//            Assert.IsNotNull(productosActivos,"No se listo bien los activos");
//            Assert.IsTrue(productosActivos.Count > 0, "La lista de productos activos no debería estar vacía.");

//            List<productoDto> productosPorTipo = productoBO.ListarPorTipo("MEDICAMENTO");
//            Assert.IsNotNull(productosPorTipo, "No se listo bien los por tipo");
//            Assert.IsTrue(productosPorTipo.Count > 0, "La lista de productos por tipo no debería estar vacía.");
//        }

//        [TestMethod]
//        public void Insertar()
//        {
//            // Crear un nuevo producto
//            productoDto nuevoProducto = new productoDto
//            {
//                tipoProducto = new tipoProductoDto { tipoProductoId = 1 },
//                nombre = "Shampoo",
//                presentacion = "Botella de 500ml",
//                precioUnitario = 15.5,
//                stock = 100,
//                activo = true
//            };

//            // Insertar el producto
//            int resultado = productoBO.Insertar(nuevoProducto);

//            // Verificar que el ID del nuevo producto sea mayor que 0 (se insertó correctamente)
//            Console.WriteLine("ID del nuevo producto: " + resultado);
//            Assert.IsTrue(resultado > 0, "El producto debería haberse insertado correctamente.");
//        }

//        [TestMethod]
//        public void Modificar()
//        {
//            // Obtener un producto existente
//            productoDto productoExistente = productoBO.ObtenerPorId(2);
//            productoExistente.nombre = "Shampoo actualizado";

//            // Modificar el producto
//            int resultado = productoBO.Modificar(
//                productoExistente.productoId,
//                productoExistente.tipoProducto.tipoProductoId,
//                productoExistente.nombre,
//                productoExistente.presentacion,
//                productoExistente.precioUnitario,
//                productoExistente.stock,
//                productoExistente.activo
//            );

//            // Verificar que la modificación fue exitosa
//            Assert.IsTrue(resultado > 0, "El producto debería haberse modificado correctamente.");
//        }

//        [TestMethod]
//        public void Eliminar()
//        {
//            // Definir el ID del producto a eliminar
//            int productoIdAEliminar = 1;

//            // Eliminar el producto
//            int resultado = productoBO.Eliminar(productoIdAEliminar);

//            // Verificar que la eliminación fue exitosa
//            Assert.IsTrue(resultado > 0, "El producto debería haberse eliminado correctamente.");
//        }

//        [TestMethod]
//        public void ObtenerPorId()
//        {
//            // Definir el ID del producto a obtener
//            int productoId = 2;

//            // Obtener el producto
//            productoDto producto = productoBO.ObtenerPorId(productoId);

//            // Verificar que el producto no sea nulo y que el ID coincida
//            Assert.IsNotNull(producto, "El producto no debería ser nulo.");
//            Assert.AreEqual(productoId, producto.productoId, "El ID del producto obtenido debería coincidir.");
//        }
//    }
//}
